palace/utils/geodata.cpp

```diff
@@ -89,55 +103,114 @@ std::unique_ptr<mfem::ParMesh> ReadMesh(MPI_Comm comm, const IoData &iodata, boo
     // the serial mesh.
-    if (reorder)
+    if (iodata.model.reorder_elements)
     {
-      ReorderMesh(*smesh);
+      ReorderMeshElements(*smesh);
+    }
+
+    // Refine the serial mesh (not typically used, prefer parallel uniform refinement
+    // instead).
+    {
+      int ne = smesh->GetNE();
+      for (int l = 0; l < iodata.model.refinement.ser_uniform_ref_levels; l++)
+      {
+        smesh->UniformRefinement();
+      }
+      if (iodata.model.refinement.ser_uniform_ref_levels > 0)
+      {
+        Mpi::Print("Serial uniform mesh refinement levels added {:d} elements (initial = "
+                   "{:d}, final = {:d})\n",
+                   smesh->GetNE() - ne, ne, smesh->GetNE());
+      }
     }
 
     // Clean up unused domain elements from the mesh, add new boundary elements for material
     // interfaces if not present. Can only clean up conforming meshes, assumes that any
     // nonconformal mesh was generated by adaptation and thus does not need checking.
-    if ((clean_elem || add_bdr) && smesh->Conforming())
+    if (smesh->Conforming())
     {
-      static_cast<void>(CheckMesh(smesh, iodata, clean_elem, add_bdr, false, nullptr));
+      static_cast<void>(CheckMesh(smesh, iodata, iodata.model.clean_unused_elements,
+                                  iodata.model.add_bdr_elements));
     }
 
     // Generate the mesh partitioning.
-    partitioning = GetMeshPartitioning(*smesh, Mpi::Size(comm), iodata.model.partition);
-
-    // Optionally add subdomain interface boundary elements.
-    if (add_subdomain && smesh->Conforming())
-    {
-      static_cast<void>(
-          CheckMesh(smesh, iodata, false, false, add_subdomain, partitioning.get()));
-    }
+    partitioning = GetMeshPartitioning(*smesh, Mpi::Size(comm), iodata.model.partitioning);
   }
 
+  // Distribute the mesh.
   std::unique_ptr<mfem::ParMesh> pmesh;
   if (use_mesh_partitioner)
   {
     pmesh = DistributeMesh(comm, smesh, partitioning.get(), iodata.problem.output);
   }
   else
   {
+    // Send the preprocessed serial mesh and partitioning as a byte string.
+    constexpr bool generate_edges = false, refine = true, fix_orientation = false;
+    std::string so;
+    int slen = 0;
+    if (smesh)
+    {
+      std::ostringstream fo(std::stringstream::out);
+      // fo << std::fixed;
+      fo << std::scientific;
+      fo.precision(MSH_FLT_PRECISION);
+      smesh->Print(fo);
+      smesh.reset();  // Root process needs to rebuild the mesh to ensure consistency with
+                      // the saved serial mesh (refinement marking, for example)
+      so = fo.str();
+      // so = zlib::CompressString(fo.str());
+      slen = static_cast<int>(so.size());
+      MFEM_VERIFY(so.size() == (std::size_t)slen, "Overflow in stringbuffer size!");
+    }
+    Mpi::Broadcast(1, &slen, 0, node_comm);
+    if (so.empty())
+    {
+      so.resize(slen);
+    }
+    Mpi::Broadcast(slen, so.data(), 0, node_comm);
+    if (!smesh)
+    {
```

Comment:
Maybe make this a `MFEM_VERIFY` the comment above about `smesh.reset()` means the logic requires it by empty so might as well enforce and then drop the if branch.

Reply:
Good catch, this is left over from when the root process was not rebuilding the mesh.